cpp/src/arrow/buffer_builder.h

```diff
@@ -350,6 +350,17 @@ class TypedBufferBuilder<bool> {
     bit_length_ += num_elements;
   }
 
+  void UnsafeAppend(const uint8_t* bitmap, int64_t offset, int64_t num_elements) {
+    if (num_elements == 0) return;
+    int64_t i = offset;
+    internal::GenerateBitsUnrolled(mutable_data(), bit_length_, num_elements, [&] {
```

Comment:
Perhaps use `CopyBitmap` instead? It should be faster on large numbers of elements.